src/Microsoft.Diagnostics.Monitoring.RestServer/Controllers/DiagController.cs

```diff
@@ -315,5 +363,69 @@ private static LogFormat ComputeLogFormat(IList<MediaTypeHeaderValue> acceptedHe
             }
             return LogFormat.None;
         }
+
+        private static ActionResult Result(
+            string providerName,
+            Func<Stream, CancellationToken, Task> action,
+            string fileName,
+            string contentType,
+            IEndpointInfo endpointInfo,
+            bool asAttachment = true)
+        {
+            if (string.IsNullOrEmpty(providerName))
+            {
+                return new OutputStreamResult(
+                    action,
+                    contentType,
+                    asAttachment ? fileName : null);
+            }
+            else
+            {
+                return new EgressStreamResult(
+                    action,
+                    providerName,
+                    fileName,
+                    endpointInfo,
+                    contentType);
+            }
+        }
+
+        private static Func<Stream, CancellationToken, Task> ConvertFastSerializeAction(Func<CancellationToken, Task<IFastSerializable>> action)
+        {
+            return async (stream, token) =>
+            {
+                IFastSerializable fastSerializable = await action(token);
+
+                // FastSerialization requests the length of the stream before serializing to the stream.
```

Comment:
I think this is a fine workaround for now, but this should really get fixed in FastSerializer if we want to do this in production. This could use a huge amount of memory.

Reply:
I added more comments and a consideration that this should be fixed in FastSerializer.